### ISSUE TITLE
Support for Wildfire maps

### ISSUE THREAD

**Issue by ja2-derek** [ja2-stracciatella/ja2-stracciatella#1011]
I was bored with the vanilla maps, so I tried playing on Wildfire maps with Straciatella, and I got a working proof-of-concept. I probably is not the first one to try, but I wonder how much interest there is. I may be able to help make this into a JA2S mod.

I am able to get in-game with WF maps, though I haven't completed a play-through nor tested extensively. I am reporting my progress so far, hoping some of you would also be interested:

## Using Wildfire maps

These are what I did so far:

- Export assets from Wildfire data directory, I am using the Steam version. Extract the below files and place in a mod directory
  - Data/BinaryData/JA2set.dat
  - Data/Interface/b_map.sti, convert to 8-bit pcx as b_map.pcx (See #712)
  - Data/Maps/**
  - Data/Radarmaps/**
  - Data/TileSets/**
- The maps should be loading fine at this point once the mod is enabled
- There are more town sectors in Wildfire maps. To enable those, edit:
  - InitializeStrategicMapSectorTownNames(void), add A1,B1,C14,F2 as town sectors
  - GARRISON_GROUP, add garrison groups to those sectors
  - g_traverse_we, g_traverse_ns and g_traverse_through, change those to town sectors
  
## Known issues

- Some sectors crashes the Map Editor. For example I6 (Estoni). I haven't been able to figure out why.
- Wildfire has different items, so you will be getting wrong pre-placed items(https://github.com/ja2-stracciatella/ja2-stracciatella/issues/321#issuecomment-221561534)
-  Some rooms and  key locations moved, which means:
   - Some NPC AI movement scripts fail, e.g. Kyle in San Mona. The NPC scripts, and any hard-coded gridNo needs to be updated with WF ones.
- Some civilians outside of San Mona are in Kingpin's faction. For example some civilians may attack you in Drassen Airport if Kingpin is hostile.

## What can be done

If there is interest, we can make it into a modpack and provide basic (and experimental) support of Vanilla-like experience on WF maps.

Regarding items differences, I don't know if we can ignore or somehow replace those WF items. I don't know how JA2 1.13 handle those issues.

For the strategic map differences, we can externalize Sector Info and Garrison Groups and make them mod-overridable.

## Screenshots

![screen1](https://user-images.githubusercontent.com/63151803/78450930-d6505080-76b4-11ea-8fa2-880b8b341da0.png)

![screen2](https://user-images.githubusercontent.com/63151803/78450931-d81a1400-76b4-11ea-8cca-dd2dfe98a6dd.png)

**Comment by lynxlynxlynx** [comment]
Cool stuff, nice investigation!

We can script the data extraction with https://github.com/ja2-stracciatella/ja2-open-toolset and other python tools if needed.

If extended, it could also take care of the item swapping. Though I don't understand the problem, are their indexes different or why do vanilla ones show up?

Point three is the easiest.

**Comment by ja2-derek** [comment]
I think new items were added in WF. Some items, such as ammos that exist in Vanilla, are working. I am yet to check if WF is only adding new items, or if it also replaces existing ones.

There are indeed WF-only items, for example this one in Drassen SAM site. 
![Items_WF](https://user-images.githubusercontent.com/63151803/78471841-cdb85280-7766-11ea-9740-7cff2eab623c.png)
In Wildfire, it gives 2 "LAW rockets". This is the "ammo" for the re-usable and reload-able LAW only in WF.


![Items_Van1](https://user-images.githubusercontent.com/63151803/78471843-cf821600-7766-11ea-9cc8-6f58f324b956.png)
In JA2S, it becomes "SMALL NAME". And the game glitches here sometimes. I have got a  crash  once, and another time seen the item appearing like a used LAW, which I suppose is some kind of default fallback.

**Comment by lynxlynxlynx** [comment]
Good, so it sounds like we'd just need to extend our externalized json tables and probably add some existence checks to the code.

**Comment by ja2-derek** [comment]
I have looked at the ITEMDESC.edt. It looks like WF added some new items, replaced some  weapons (e.g. Rocket Rifle removed, Colt .45 replaced with UMP). Keys and other plot-significant items seem to be unchanged.

![items_diff3](https://user-images.githubusercontent.com/63151803/78509792-4a791a00-77c3-11ea-81e4-3370afaa6358.png)
I am doing all these with the Steam version of WF (It says v6.04).

Do you know of any tools that can export .EDT as text or json, so I can produce a full diff?

**Comment by lynxlynxlynx** [comment]
Seven said: EdtMegaEditor

**Comment by derek9gag** [comment]
I made the screencap diff using EdtMegaEditor! Now I see it allows me to save as .TXT. Thank you.

**Comment by ja2-derek** [comment]
I have been using EdtMegaEditor, but I didn't realized I can save as .TXT 🤦‍♂ 

Anyway, full diff below. WF has only used a few unused slots (for new weapons), and replaced ones are mostly weapon items too. So if JA2S can ignore undefined items, it should not break gameplay too much. But of course, game balance will be a different issue.

<details>
<summary>
Expand to see the diff between Vanilla("<") and Wildfire(">"))
</summary>

```diff
2,7c2,7
< 1,Glock 17
< 2,Glock 18
< 3,Beretta 92F
< 4,Beretta 93R
< 5,.38 S&W
< 6,Barracuda
---
> 1,Calico M950
> 2,Micro Uzi
> 3,USP
> 4,Beretta 92F
> 5,Colt 1991
> 6,Colt Python
9c9
< 8,Colt .45
---
> 8,UMP
12,15c12,15
< 11,Thompson
< 12,Commando
< 13,MP53
< 14,AKSU-74
---
> 11,Uzi
> 12,M4
> 13,MP5
> 14,Colt Commando
17,20c17,20
< 16,Type-85
< 17,SKS
< 18,Dragunov
< 19,M24
---
> 16,HK 53
> 17,Mini-30
> 18,SVD
> 19,M16
22c22
< 21,G41
---
> 21,HK 33
24,25c24,25
< 23,C-7
< 24,FA-MAS
---
> 23,G36
> 24,MSG90
28,31c28,31
< 27,M-14
< 28,FN-FAL
< 29,G3A3
< 30,G11
---
> 27,M14
> 28,FAL
> 29,G3
> 30,AN-94
36,37c36,37
< 35,RPK-74
< 36,H&K 21
---
> 35,RPK
> 36,HK 21
56,57c56,57
< 55,Rocket Rifle
< 56,Automag
---
> 55,SVU
> 56,MP7
66,69c66,69
< 65,A. R. Rifle
< 66,'small name'
< 67,'small name'
< 68,'small name'
---
> 65,VAL
> 66,VSS
> 67,V-94
> 68,F2000
72,80c72,80
< 71,9mm mag
< 72,9mm SMG mag
< 73,9mm mag AP
< 74,9mm SMG mag AP
< 75,9mm mag HP
< 76,9mm SMG mag HP
< 77,.38 S-load
< 78,.38 S-load AP
< 79,.38 S-load HP
---
> 71,9mm mag AP
> 72,9mm SMG mag AP
> 73,9mm mag AP 50
> 74,9mm mag HP
> 75,9mm SMG mag HP
> 76,9mm mag HP 50
> 77,4.6mm mag
> 78,7.62*54mm 10 AP
> 79,7.62*54mm 10 HP
87,106c87,106
< 86,.357 S-load
< 87,.357 mag
< 88,.357 S-load AP
< 89,.357 mag AP
< 90,.357 S-load HP
< 91,.357 mag HP
< 92,5.45mm mag
< 93,5.45mm mag HP
< 94,5.56mm mag
< 95,5.56mm mag HP
< 96,7.62mm WP 10
< 97,7.62mm WP 30
< 98,7.62mm WP 10
< 99,7.62mm WP 30
< 100,7.62mm NATO 5
< 101,7.62mm NATO 20
< 102,7.62mm NATO 5
< 103,7.62mm NATO 20
< 104,4.7mm mag
< 105,5.7mm mag
---
> 86,.357 S-load AP
> 87,.357 mag AP
> 88,.357 S-load HP
> 89,.357 mag HP
> 90,5.45mm mag
> 91,5.45mm mag HP
> 92,5.56mm mag
> 93,5.56mm 100 AP
> 94,5.56mm mag HP
> 95,5.56mm 100 HP
> 96,7.62mm WP 30 AP
> 97,7.62mm WP 75 AP
> 98,7.62mm WP 30 HP
> 99,7.62mm WP 75 HP
> 100,7.62mm NATO 20
> 101,7.62mm NATO 100
> 102,7.62mm NATO 20
> 103,7.62mm NATO 100
> 104,9*39mm 10
> 105,9*39mm 20
112,114c112,114
< 111,Minirockets, AP
< 112,Minirockets, HE
< 113,Minirockets, HEAP
---
> 111,12.7mm, AP
> 112,12.7mm, HE
> 113,12.7mm, HEAP
161c161
< 160,'small name'
---
> 160,LAW rocket
203c203
< 202,Med Kit
---
> 202,Hospital
208c208
< 207,Silencer
---
> 207,Silencer 9mm
248c248
< 247,Discarded LAW
---
> 247,Empty LAW
266c266
< 265,Golf Clubs
---
> 265,Silencer .45
270c270
< 269,Cigars
---
> 269,Silencer P90
306,308c306,308
< 305,Rod
< 306,Spring
< 307,Rod & Spring
---
> 305,Silencer MP7
> 306,Silencer Commando
> 307,Collimator
```

</details>

**Comment by ja2-derek** [comment]
Implemented the installation script. You can try Wildfire Maps using this:

https://github.com/ja2-derek/ja2-stracciatella/tree/wildfire-maps-mod/assets/mods/wildfire-maps

**Comment by lynxlynxlynx** [comment]
Wow, nice job! If you're interested in finishing this, I suggest we keep this bug as the meta overview one and you open new subreports as needed. 

The city sector externalizatoin is the easy part. 
Weapons we already have the same way (eg. assets/externalized/weapons.json), so it's mostly a matter of checking if any new hardcoded needs are there or changed, then adapting the externalized files (I guess also the merchant inventories).
If it's not the case yet, ensuring mod data overrides these files as well, since it's likely we can't ship it by default (ok, it could be a separate task to generalize this out).

Also, if you want, we can move the repo under this github organisation.

**Comment by ja2-derek** [comment]
There are indeed quite a lot more to work on, to get this fully working. I am going to use this comment to track as you said. I will come back and update this as I go deeper. The externalization should benefit the project anyways.

I am using this comment to track the work breakdown:

## Data files exports from Wildfire

- [x] Data/BinaryData/JA2set.dat
- [x] Data/Interface/b_map.sti, convert as b_map.pcx
- [x] Data/Interface/MilitiaMaps.sti
- [x] Data/Maps/**
- [ ] Data/Npcdata/, requires patching to fix NPC pathing
  - [ ] 095.npc, Kyle - his Tattoo shop has moved
- [x] Data/Radarmaps/**
- [x] Data/TileSets/**

## Data in codebase to be externalized 

With the suggested name and location in codebase:

- [ ] `strategic-map-base-sectors.json` - Map_Screen_Interface_Map.cc: sBaseSectorList[]; update Chitzena, Grumm for militia maps
- [ ] `strategic-map-town-names.json` - StrategicMap.cc: InitializeStrategicMapSectorTownNames(void); add A1,B1,C14,F2 as town sectors
- [ ] `strategic-map-movement-costs.json` - Strategic_Movement_Costs.cc: g_traverse_we, g_traverse_ns and g_traverse_through; change those to town sectors
- [ ] `strategic-ai-garrison-groups.json` - StrategicAI.cc: gOrigGarrisonGroup[]; add garrison groups to those sectors
- [ ] `npc-action-gridnos.json` - Interface_Dialogue.cc: HandleNPCDoAction() 
    - [ ] NPC_ACTION_KYLE_GETS_MONEY: update sGridNo, corresponding to Data/Npcdata/095.npc
- [ ] `strategic-map-bloodcat-sectors.json` - Game_Init.cc: InitBloodCatSectors
- [ ] `strategic-map-vehicle-placements.json` - Strategic_Event_Handler::HandleEarlyMorningEvents
      - Missing vehicle placement (Hamous and his truck) at D3. Vehicle only in alternate version of G6_a.dat, but there is no trigger to activate it

## Other issues to be addressed

- [ ] Handling of pre-placed Wildfire-only items
- [ ] I6 crashes the editor and game

**Comment by lynxlynxlynx** [comment]
I don't have WF, so can only pontificate. What don't I understand about 095.npc? Doesn't just replacing the file work?

**Comment by ja2-derek** [comment]
> I don't have WF, so can only pontificate. What don't I understand about 095.npc? Doesn't just replacing the file work?

Oh yes. How come I didn't think of that?

**Comment by ja2-derek** [comment]
OK, the packaging is done for the installer. https://github.com/ja2-derek/mod-wildfire-maps/releases/tag/0.1.0

**Comment by lynxlynxlynx** [comment]
Nice. Did you consider the installer taking an optional path to WF, so it could copy slfs itself?

**Comment by ja2-derek** [comment]
Installer updated: https://github.com/ja2-derek/mod-wildfire-maps/releases/tag/0.2.0

 - Installer now accepts `--src_dir` command arg, or launches a GUI prompt if display is available
 - Installer output are now highlighted with colors
 - Excludes item sprites from Tileset, so items will look correct in Tactical
 - Includes all the externalized JSON assets so far. Already customized for WF maps, but still need further data verification

**Comment by lynxlynxlynx** [comment]
Cool. Looks like just a bit more is needed to obviate the need for point 3 in the README.

**Comment by lynxlynxlynx** [comment]
What's with the other npc files? I thought that just preserving the version from Wildfire is enough and we don't need to care.

**Comment by ja2-derek** [comment]
I am hand-picking which NPC files to use. It is because some NPCs have dependencies on WF version off quotes, speeches or code logic. I want to keep this mod to only maps if possible.

So by default I am using the vanilla NPC. If there are path changes, use the Wildfire version. There is one exception case KYLE, which I am patching the vanilla file to work with the WF map

**Comment by lynxlynxlynx** [comment]
So I guess that's blocked by the interpreter discussion, ok. Can you then verify the other grid todos before jumping to UB?

**Comment by ja2-derek** [comment]
Yes I am still on it. I am going to finish a playthrough with WF maps, and verify each and every one of the Grid TODOs.

If there are no blocking issues with GridNos, I will update the installer.  Then it should be ready for release targeting 0.17.

**Comment by ja2-derek** [comment]
### Current status

- No outstanding blocker issues
- Installer script ([mod-wildfire-maps](https://github.com/ja2-derek/mod-wildfire-mapshttps://github.com/ja2-derek/mod-wildfire-maps)) updated, ready for 0.17 release
- Completed played through, covered all significant parts that I can think of.
- A couple of open non-blocker issues: 
    - Creature lair externalization is half done and not tested.  I don't have time to finish that just now
    - I don't really want to do  the "tactical gridNo" one. I'd rather wait for scripted / externalized quest logic

**Comment by ja2-derek** [comment]
I'm going to close this issue, as all of the necessary changes are merged.

Creature Lair externalization will be handled with #1087.

Tactical gridNo stuff will be deferred

**Comment by ja2-derek** [comment]
Latest version of installer script: https://github.com/ja2-derek/mod-wildfire-maps/releases/tag/0.2.1

**Comment by lynxlynxlynx** [comment]
Re issue: yeah, please open a new one and migrate what's left to do for a full WF experience.

Repo: yes, please. Do you have the option already?

I would like to invite you to become part of the github team — it's well deserved and might be needed for the migration. @flaviojs any objections?

**Comment by flaviojs** [comment]
No objections. 👍

**Comment by lynxlynxlynx** [comment]
Ok, done. Welcome aboard! :fireworks: 

In case it wasn't obvious @ja2-derek, we all work on the main repo indirectly through PRs (master branch is marked as protected).

I'll write something for the website features page as soon as you move the repo, so the links won't break.

**Comment by ja2-derek** [comment]
Thank you! 

Moved repo: https://github.com/ja2-stracciatella/mod-wildfire-maps

**Comment by lynxlynxlynx** [comment]
Done on my side as well. So then please just open the new issue and this one is done done.

https://ja2-stracciatella.github.io/features/#bundled-optional-mini-mods
